### PR TITLE
[POC] extend symfony FormBuilder with methods addBefore & addAfter

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | POC. Introduce methods addBefore & addAfter in FormBuilder, these would save a lot of time for module developers. 
| Type?             | improvement
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | yes
| Fixed ticket?     | 
| How to test?      | Will define later if this pr reaches that state (its only a playground now)
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24337)

### CODE REVIEW

**Comment by zuk3975** [comment]
I've extended the symfony form builder, but not sure how to proceed. FormFactoryInterface is used in every form, so I could just add new implementation for it, but I then the return type won't be compatable with the interface?
Anyway, the whole idea should be clear by these couple code lines, need help to finish up

File: src/PrestaShopBundle/Form/FormBuilder.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShopBundle\Form;
+
+use RuntimeException;
+
+final class FormBuilder extends \Symfony\Component\Form\FormBuilder implements FormBuilderInterface
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/FormBuilder.php:32]: Remove `final`
Move `FormBuilder` to use section

File: src/PrestaShopBundle/Form/FormBuilder.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+{
+    public function addBefore(string $targetFieldName, $child, $type = null, array $options = []): FormBuilderInterface
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/FormBuilder.php:34]: Could you typehint parameters ?

File: src/PrestaShopBundle/Form/FormBuilder.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+            if ($formBuilder->getName() === $targetFieldName) {
+                $this->add($child, $type, $options);
+            }
+        }
+
+        return $this;
+    }
+
+    public function addAfter(string $targetFieldName, $child, $type = null, array $options = []): FormBuilderInterface
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/FormBuilder.php:50]: Could you typehint parameters ?

File: src/PrestaShopBundle/Form/FormBuilderInterface.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShopBundle\Form;
+
+interface FormBuilderInterface extends \Symfony\Component\Form\FormBuilderInterface
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/FormBuilderInterface.php:31]: Could you move `\Symfony\Component\Form\FormBuilderInterface` to use section ?

**Review by Progi1984 — COMMENTED** [review]
Need unit tests

File: src/PrestaShopBundle/Form/FormBuilder.php (hunk 1 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/FormBuilder.php:32]: Indeed something like:
```suggestion
use Symfony\Component\Form\FormBuilder as BaseFormBuilder;

class FormBuilder extends BaseFormBuilder implements FormBuilderInterface
```

File: src/PrestaShopBundle/Form/FormBuilder.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace PrestaShopBundle\Form;
+
+use RuntimeException;
+
+final class FormBuilder extends \Symfony\Component\Form\FormBuilder implements FormBuilderInterface
+{
+    public function addBefore(string $targetFieldName, $child, $type = null, array $options = []): FormBuilderInterface
+    {
+        $this->assertFieldExists($targetFieldName);
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/FormBuilder.php:36]: We have a component https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/PrestaShopBundle/Form/FormBuilderModifier.php which does the same but "externally" I believe I copied it from one of your PR in the example module

It's been merged in 178 to ease the module developers extend our form I think it should be used internally in this class by injecting it

Comment (jolelievre) [reply src/PrestaShopBundle/Form/FormBuilder.php:36]:
```suggestion
        $this->modifier->addAfter($this, $targetFieldName, $newChild, $type, $options);
        
        return $this;
```

**Review by jolelievre — COMMENTED** [review]
This POC is a really good idea But I think it would be even greater if all the form builder were automatically using this class To do this we need to override the Symfony form factory system to put our classes in the middle

I researched this a lot in this PR https://github.com/PrestaShop/PrestaShop/pull/22838 it's not intuitive at all but it's doable and it should be a bit simpler than the use case I implemented, there are a few layers that need to be overridden:

- create a subclass a `ResolvedFormType` which overrides protected function `newBuilder` to use our own
- create a `ResolvedFormTypeFactoryInterface` factory which decorates the framework one `form.resolved_type_factory`

**Comment by zuk3975** [comment]
I think its done already by some other PR.

**Pull request closed without merging**